woocommerce/woocommerce-gateway-stripe
# Add Docker E2E environment setup and runner

Fixes #2576, #2530

## Changes proposed in this Pull Request:
This PR adds an E2E test workflow for open PRs.

The `npm run test:e2e-setup` and `npm run test:e2e` scripts were extended to support not receiving a `base_url` parameter.

When no `base_url` param is present, an independent E2E docker environment is configured/used for the tests, which allows the tests to run in a GitHub Workflow on each Pull Request.

## Testing instructions

### Verify that **all** PR checks pass
Including the new `E2E tests / WP=latest, WC=latest, PHP=7.4 (pull_request)`:

![Screenshot 2024-02-21 at 12 44 27](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/7f6325ba-cb4c-4884-90d9-ab680b051893)


### Test the Jurassic Ninja runner:
1. Create a JN site
2. Update the credentials for the newly created JN site in your `tests/e2e/config/local.env`
3. Run the E2E suite with the JN base url:
    `npm run test:e2e-setup -- --base_url=<JN_URL>`
4. Verify that the setup and all 22 e2e tests succeed
    ![Screenshot 2024-02-21 at 12 43 42](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/0a162bc2-8b5b-4e4c-b0b0-dcae2a9c5538)


### Test the docker runner:
0. Build the extensions asses
    `npm run build`

1. Initialize/start the E2E docker containers:
    `npm run test:e2e-setup`

2. Verify that the environment was successfully started: 
    ![Screenshot 2024-02-21 at 12 32 29](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/ee27fac8-0ab8-48f6-a610-888f5f5fac3f)

3. Run the E2E suite (without the `base_url` param):
    `npm run test:e2e`

6. Verify that the setup and all 22 e2e tests succeed:
    ![Screenshot 2024-02-21 at 12 35 04](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/685833c8-c979-4daa-b2a6-b771af45438f)

7. Stop the E2E docker containers:
    `npm run test:e2e-down`
